tracer/test/Datadog.Trace.Security.Unit.Tests/ObjectExtractorTests.cs

```diff
@@ -130,13 +121,49 @@ public void TestAnonymousType()
             var result = ObjectExtractor.Extract(target) as Dictionary<string, object>;
 
             Assert.NotNull(result);
-
             Assert.Equal(target.Dog1, result[nameof(target.Dog1)]?.ToString());
             Assert.Equal(target.Dog2, result[nameof(target.Dog2)]?.ToString());
             result[nameof(target.Id)].Should().BeOfType<int>();
             target.Id.Should().Be((int)result[nameof(target.Id)]);
         }
 
+        [Fact]
+        public void TestAnonymousTypeEmpty()
+        {
+            var target = new { };
+
+            var result = ObjectExtractor.Extract(target) as Dictionary<string, object>;
+            Assert.NotNull(result);
+            Assert.Empty(result);
+        }
+
+        [Fact]
+        public void TestAnonymousTypeNested()
+        {
+            var target = new { Name = "Outer", Dog = new { Name = "Inner" }, Cat = new { } };
+
+            var result = ObjectExtractor.Extract(target) as Dictionary<string, object>;
+            Assert.NotNull(result);
+            result.Should().HaveCount(3);
+            result.Should().HaveElementAt(0, new KeyValuePair<string, object>("Name", "Outer"));
+            result.ElementAt(1).Key.Should().Be("Dog");
+            result.ElementAt(1).Value.Should().BeEquivalentTo(new Dictionary<string, object> { { "Name", "Inner" } });
+            result.ElementAt(2).Key.Should().Be("Cat");
+            result.ElementAt(2).Value.Should().BeEquivalentTo(new Dictionary<string, object>(0));
+        }
+
+        [Fact]
+        public void TestAnonymousTypeArray()
+        {
+            var target = new[] { new { Name = "Anon1" }, new { Name = "Anon2" }, new { Name = "Anon1" } };
+            var result = ObjectExtractor.Extract(target) as List<object>;
+            result.Should().NotBeNull();
+            result.Should().HaveCount(3);
+            result!.ElementAt(0).Should().BeEquivalentTo(new Dictionary<string, object> { { "Name", "Anon1" } });
+            result.ElementAt(1).Should().BeEquivalentTo(new Dictionary<string, object> { { "Name", "Anon2" } });
+            result.ElementAt(2).Should().BeEquivalentTo(new Dictionary<string, object>(0));
```

Comment:
Interesting, so it de-dupes here? I didn't expect that! 😃

Reply:
Yes [it does here](https://github.com/DataDog/dd-trace-dotnet/blob/43efa46a49669568b3bd2d6aca6675ae03be4d19/tracer/src/Datadog.Trace/AppSec/ObjectExtractor.cs#L52), which isn't so bad as the waf would have to de duplicate anyway further down the line